packages/sui-gateway/src/base_types.rs

```diff
@@ -0,0 +1,94 @@
+use std::str::FromStr;
+
+use error_stack::{report, Report, Result, ResultExt};
+use serde::{Deserialize, Serialize};
+
+use crate::error::Error;
+
+const ADDRESS_PREFIX: &str = "0x";
+const SUI_ADDRESS_LENGTH: usize = 32;
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct SuiAddress([u8; SUI_ADDRESS_LENGTH]);
+
+impl SuiAddress {
+    pub fn as_bytes(&self) -> &[u8; SUI_ADDRESS_LENGTH] {
+        &self.0
+    }
+}
+
+impl TryFrom<&[u8]> for SuiAddress {
+    type Error = Report<Error>;
+
+    fn try_from(bytes: &[u8]) -> Result<Self, Error> {
+        bytes
+            .try_into()
+            .map(Self)
+            .change_context(Error::InvalidAddressBytes(bytes.to_vec()))
+    }
+}
+
+impl FromStr for SuiAddress {
+    type Err = Report<Error>;
+
+    fn from_str(s: &str) -> Result<Self, Error> {
+        hex::decode(
+            s.strip_prefix(ADDRESS_PREFIX)
+                .ok_or(report!(Error::InvalidAddressHex(s.to_string())))?,
+        )
+        .change_context(Error::InvalidAddressHex(s.to_string()))?
+        .as_slice()
+        .try_into()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use rand::RngCore;
+
+    use super::*;
+
+    #[test]
+    fn test_sui_address_try_from() {
```

Comment:
test_ prefix isn't needed i think
```suggestion
    fn sui_address_try_from() {
```